metricflow/query/group_by_item/candidate_push_down/push_down_visitor.py

```diff
@@ -0,0 +1,408 @@
+from metricflow.specs.patterns.none_date_part import NoneDatePartPattern
+from metricflow.specs.patterns.spec_pattern import SpecPattern
+from metricflow.specs.specs import InstanceSpecSet, LinkableInstanceSpec
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class PushDownResult:
+    """The result that is pushed down from the root nodes to the leaf nodes to resolve ambiguous group-by items."""
+
+    # The set of candidate specs that could match the ambiguous group-by item.
+    candidate_set: GroupByItemCandidateSet
+    # The issues seen so far while pushing down the result / resolving the ambiguity.
+    issue_set: MetricFlowQueryResolutionIssueSet
+
+    def __post_init__(self) -> None:  # noqa: D
+        # If there are no errors, there should be a candidate spec in each candidate set.
+        # If there are errors, there shouldn't be any candidate sets.
+        assert (not self.issue_set.has_errors and not self.candidate_set.is_empty) or (
+            self.issue_set.has_errors and self.candidate_set.is_empty
+        ), mf_pformat_many(
+            "candidate_set / issue_set mismatch:", {"candidate_set": self.candidate_set, "issue_set": self.issue_set}
+        )
+
+    def filter_candidates_by_pattern(self, spec_pattern: SpecPattern) -> PushDownResult:
+        """Create a new result where only the candidates matching the pattern are included."""
+        return PushDownResult(
+            candidate_set=self.candidate_set.filter_candidates_by_pattern(spec_pattern),
+            issue_set=self.issue_set,
+        )
+
+    def filter_candidates_by_patterns(self, spec_patterns: Sequence[SpecPattern]) -> PushDownResult:
+        """Create a new result where only the candidates matching the patterns are included."""
+        if len(spec_patterns) == 0:
+            return self
+
+        candidate_set = self.candidate_set
+        for spec_pattern in spec_patterns:
+            candidate_set = candidate_set.filter_candidates_by_pattern(spec_pattern)
+
+        return PushDownResult(
+            candidate_set=candidate_set,
+            issue_set=self.issue_set,
+        )
+
+
+class DagTraversalPathTracker:
+    """Helps track the path traveled by a visitor through the nodes in a group-by-tem resolution DAG."""
+
+    def __init__(self) -> None:  # noqa: D
+        self._current_path: List[GroupByItemResolutionNode] = []
+
+    @contextmanager
+    def track_node_visit(self, node: GroupByItemResolutionNode) -> Iterator[MetricFlowQueryResolutionPath]:
+        """Context manager used to wrap the visit calls.
+
+        e.g.
+
+            def visit_some_node(self, node) -> Result:
+                with track_node_visit(node) as current_traversal_path:
+                    # current_traversal_path contain the nodes visited from the start.
+        """
+        self._current_path.append(node)
+        yield MetricFlowQueryResolutionPath(tuple(self._current_path))
+        self._current_path.pop(-1)
```

Comment:
Maybe add a bit to the docblock to illustrate the need for this last line, because on its own it looks like this doesn't maintain state. Something like:

```
def visit_some_node(self, node) -> Result:
    with track_node_visit(node as current_traversal_path):
        # current_traversal_path contain the nodes visited before this
        # do things

    # context manager removes this node from current_traversal_path so sibling branches don't include it
```